File: Sources/Core/AZAppDelegate.m
#import "AZAppDelegate.h"


@implementation AZAppDelegate {

	UIWindow *window;

}

// can't use self before the app delegate is initialized so I can't make ivars :frSmh:
static Class strongClass;
static UIButton *quitButton;
static UILabel *addressLabel;
static UIWindow *strongWindow;
static AuthManager *authManager;

#define kAzureLilacTintColor [UIColor colorWithRed: 0.70 green: 0.56 blue: 1.0 alpha: 1.0]
#define kAzureMintTintColor [UIColor colorWithRed: 0.40 green: 0.81 blue: 0.73 alpha: 1.0]

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {

	allocateClass(strongClass);

	authManager = [AuthManager new];
	window = [UIWindow new];
	window.tintColor = kAzureLilacTintColor;
	window.backgroundColor = UIColor.systemBackgroundColor;
	window.rootViewController = [strongClass new];
	[window makeKeyAndVisible];

	strongWindow = window;

	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	BOOL usesBiometrics = [defaults boolForKey: @"useBiometrics"];
	if(usesBiometrics && [authManager shouldUseBiometrics]) unsafePortalDispatch();
	else window.rootViewController = [AzureRootVC new];

	UINavigationBar.appearance.shadowImage = [UIImage new];
	UINavigationBar.appearance.translucent = NO;
	UINavigationBar.appearance.barTintColor = UIColor.systemBackgroundColor;

	return YES;

}

static void allocateClass(Class theClass) {

	theClass = strongClass;
	strongClass = objc_allocateClassPair([UIViewController class], "NotAuthenticatedVC", 0);
	Method vDLMethod = class_getInstanceMethod([UIViewController class], @selector(viewDidLoad));
	const char *types = method_getTypeEncoding(vDLMethod);
	class_addMethod(
		strongClass,
		@selector(viewDidLoad),
		(IMP) overrideVDL,
		types
	);
	objc_registerClassPair(strongClass);

}

static void overrideVDL(UIViewController *self, SEL _cmd) {

	// Do any additional setup after loading the view, typically from a nib.

    /*--- idfk how to call super at runtime yet so I'll leave it like this for now,
    but it doesn't really matter anyways, since I just want a plain vanilla view controller ---*/

    // edit: I found out how, yet what I tried crashes so far, hmmm

	self.view.backgroundColor = UIColor.systemBackgroundColor;

	addressLabel = [UILabel new];
	addressLabel.alpha = 0;
	addressLabel.font = [UIFont systemFontOfSize: 12];
	addressLabel.text = [NSString stringWithFormat: @"%p", &self];
	addressLabel.transform = CGAffineTransformMakeScale(0.1, 0.1);
	addressLabel.textColor = UIColor.systemGrayColor;
	addressLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview: addressLabel];

	[addressLabel.centerXAnchor constraintEqualToAnchor: self.view.centerXAnchor].active = YES;
	[addressLabel.bottomAnchor constraintEqualToAnchor: self.view.safeAreaLayoutGuide.bottomAnchor constant: -30].active = YES;

	quitButton = [UIButton new];
	quitButton.alpha = 0;
	quitButton.transform = CGAffineTransformMakeScale(0.1, 0.1);
	quitButton.backgroundColor = kAzureMintTintColor;
	quitButton.layer.cornerCurve = kCACornerCurveContinuous;
	quitButton.layer.cornerRadius = 20;
	quitButton.translatesAutoresizingMaskIntoConstraints = NO;
	[quitButton setTitle:@"Quit" forState: UIControlStateNormal];
	[quitButton addTarget:(AZAppDelegate *)(UIApplication.sharedApplication.delegate) action:@selector(didTapQuitButton) forControlEvents: UIControlEventTouchUpInside];
	[self.view addSubview: quitButton];

	[quitButton.centerXAnchor constraintEqualToAnchor: self.view.centerXAnchor].active = YES;
	[quitButton.centerYAnchor constraintEqualToAnchor: self.view.centerYAnchor].active = YES;
	[quitButton.widthAnchor constraintEqualToConstant: 120].active = YES;
	[quitButton.heightAnchor constraintEqualToConstant: 40].active = YES;

}

static void unsafePortalDispatch() {

	[authManager setupAuthWithReason:@"Azure needs you to authenticate in order to access the app."
		reply:^(BOOL success, NSError *error) {

			dispatch_async(dispatch_get_main_queue(), ^{

				if(!success && error.code != -5) {

					strongWindow.rootViewController = [strongClass new];

					[UIView animateWithDuration:0.5 delay:0.8 options:UIViewAnimationOptionCurveEaseIn animations:^{

						addressLabel.alpha = 0.5;
						quitButton.alpha = 1;
						quitButton.transform = CGAffineTransformMakeScale(1, 1);
						addressLabel.transform = CGAffineTransformMakeScale(1, 1);

					} completion: nil];

				}

				else strongWindow.rootViewController = [AzureRootVC new];

			});

		}

	];

}


- (void)didTapQuitButton { abort(); }

@end

File: Sources/Core/AZAppDelegate.swift
import UIKit
import LocalAuthentication
import ObjectiveC.runtime


@UIApplicationMain
final class AZAppDelegate: UIResponder, UIApplicationDelegate {

	var window: UIWindow?
	private var strongWindow: UIWindow!

	private static var NotAuthenticatedVClass: AnyClass!
	private lazy var NotAuthenticatedVC = AZAppDelegate.NotAuthenticatedVClass.alloc() as? UIViewController

	private let authManager = AuthManager()

	func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
		setupNotAuthenticatedVC()

		window = UIWindow()
		window?.tintColor = .kAzureLilacTintColor
		window?.backgroundColor = .systemBackground
		window?.rootViewController = UIViewController()
		window?.makeKeyAndVisible()

		strongWindow = window

		UINavigationBar.appearance().shadowImage = UIImage()
		UINavigationBar.appearance().setBackgroundImage(UIImage(), for: .default)

		let usesBiometrics = UserDefaults.standard.bool(forKey: "useBiometrics")
		if usesBiometrics && authManager.shouldUseBiometrics() { unsafePortalDispatch() }
		else { window?.rootViewController = AzureRootVC() }

		return true
	}

	private func unsafePortalDispatch() {
		authManager.setupAuth(withReason: .unlockApp, reply: { [weak self] success, error in
			DispatchQueue.main.async {
				let laError = error as? LAError
				guard success && laError?.code != .passcodeNotSet else {
					self?.strongWindow.rootViewController = self?.NotAuthenticatedVC
					return
				}
				self?.strongWindow.rootViewController = AzureRootVC()
			}
		})
	}

	private func setupNotAuthenticatedVC() {
		allocateClass { notAuthenticatedVC in
			AZAppDelegate.sendSuper()

			let quitButton = UIButton()
			quitButton.alpha = 0
			quitButton.transform = .init(scaleX: 0.1, y: 0.1)
			quitButton.backgroundColor = .kAzureMintTintColor
			quitButton.layer.cornerCurve = .continuous
			quitButton.layer.cornerRadius = 20
			quitButton.setTitle("Retry", for: .normal)
			quitButton.addTarget(notAuthenticatedVC, action: NSSelectorFromString("didTapRetryButton"), for: .touchUpInside)

			notAuthenticatedVC.view.addSubview(quitButton)
			notAuthenticatedVC.view.centerViewOnBothAxes(quitButton)
			notAuthenticatedVC.view.setupSizeConstraints(forView: quitButton, width: 120, height: 40)

			UIView.animate(withDuration: 0.5, delay: 0.5, options: .curveEaseIn) {
				quitButton.alpha = 1
				quitButton.transform = .init(scaleX: 1, y: 1)
			}

			AZAppDelegate.didTapRetryButton { self.unsafePortalDispatch() }
		}
	}

	private func allocateClass(imp: @escaping @convention(block) (_ self: UIViewController) -> ()) {
		AZAppDelegate.NotAuthenticatedVClass = objc_allocateClassPair(UIViewController.self, "NotAuthenticatedVC", 0)!

		let method = class_getInstanceMethod(UIViewController.self, #selector(UIViewController.viewDidLoad))!
		let azure_viewDidLoad = imp_implementationWithBlock(unsafeBitCast(imp, to: AnyObject.self))
		let typeEncoding = method_getTypeEncoding(method)!
		class_addMethod(AZAppDelegate.NotAuthenticatedVClass, #selector(UIViewController.viewDidLoad), azure_viewDidLoad, typeEncoding)

		objc_registerClassPair(AZAppDelegate.NotAuthenticatedVClass)
	}

	private static func sendSuper() {
		let superclass: AnyClass = class_getSuperclass(NotAuthenticatedVClass)!
		let selector = #selector(UIViewController.viewDidLoad)
		let imp = class_getMethodImplementation(superclass, selector)

		typealias ObjcVoidFunc = @convention(c) (AnyObject, Selector) -> ()

		let superCall = unsafeBitCast(imp, to: ObjcVoidFunc.self)
		superCall(NotAuthenticatedVClass, selector)
	}

	private static func didTapRetryButton(imp: @escaping @convention(block) () -> ()) {
		class_addMethod(
			NotAuthenticatedVClass,
			NSSelectorFromString("didTapRetryButton"),
			imp_implementationWithBlock(unsafeBitCast(imp, to: AnyObject.self)),
			"v@:"
		)
	}

}
